libz-rs-sys/src/gz.rs

```diff
@@ -1296,6 +1303,340 @@ unsafe fn gz_decomp(state: &mut GzState) -> Result<(), ()> {
     Ok(())
 }
 
+/// Compress and write the len uncompressed bytes at buf to file.
+///
+/// # Returns
+///
+/// - The number of uncompress bytes written, on success.
+/// - Or 0 in case of error.
+///
+/// # Safety
+///
+/// - `file`, if non-null, must be an open file handle obtained from [`gzopen`] or [`gzdopen`].
+/// - `buf` must point to at least `len` bytes of readable memory.
+#[cfg_attr(feature = "export-symbols", export_name = crate::prefix!(gzwrite))]
+pub unsafe extern "C-unwind" fn gzwrite(file: gzFile, buf: *const c_void, len: c_uint) -> c_int {
+    let Some(state) = (unsafe { file.cast::<GzState>().as_mut() }) else {
+        return 0;
+    };
+
+    // Check that we're writing and that there's no error.
+    if state.mode != GZ_WRITE || state.err != Z_OK {
+        return 0;
+    }
+
+    // Check that the requested number of bytes can be represented by the return type.
+    if c_int::try_from(len).is_err() {
+        // Safety: we confirmed above that state is valid.
+        unsafe {
+            gz_error(
+                state,
+                Some((Z_DATA_ERROR, "requested length does not fit in int")),
+            )
+        };
+        return 0;
+    }
+
+    // Safety: We validated state above, and the caller is responsible for ensuring
+    // that buf points to at least len bytes of readable memory.
+    unsafe { gz_write(state, buf, len) }
+}
+
+// Internal implementation of `gzwrite`.
+//
+// # Returns
+// - The number of uncompress bytes written, on success.
+// - Or 0 in case of error.
+//
+// # Safety
+//
+/// - `state` must have been properly initialized, e.g. by [`gzopen_help`].
+/// - `buf` must point to at least `len` bytes of readable memory.
+/// - The caller must ensure that `len` fits in a `c_int`.
+unsafe fn gz_write(state: &mut GzState, mut buf: *const c_void, mut len: c_uint) -> c_int {
```

Comment:
the C code uses `size_t len`, that's not equivalent to `c_uint` right?

Reply:
Right, I'll change this to `usize`.

Reply:
Interestingly, there's a problem with using `usize`. As far as I can tell, a C `unsigned int` is allowed to be bigger than a C pointer, in case someone ever builds a computer with, for example, a 32-bit address space and a 128-bit ALU. So this code path:
```
pub unsafe extern "C-unwind" fn gzwrite(file: gzFile, buf: *const c_void, len: c_uint) -> c_int {
    ...
    gz_write(state, buf, len as usize)
}
```
could lead to a silent truncation of the `len` value when casting to `usize`. The safety preconditions I've documented for `gzwrite` include "`buf` must point to at least `len` bytes of readable memory," which ensures that `len` will fit in a `usize`. But I'll also use a checked cast here for extra future-proofing.